File: python/tests/test_config.py
"""Tests for configuration validation"""

import pytest
import sys
import tempfile
from pathlib import Path

from feedpulse.config import validate_url, validate_feed, validate_settings, ConfigError


def test_validate_url():
    """Test URL validation"""
    assert validate_url("https://example.com") == True
    assert validate_url("http://example.com") == True
    assert validate_url("https://example.com/path?query=1") == True
    
    assert validate_url("ftp://example.com") == False
    assert validate_url("not-a-url") == False
    assert validate_url("") == False


def test_validate_feed_success():
    """Test successful feed validation"""
    feed_data = {
        'name': 'Test Feed',
        'url': 'https://example.com/feed',
        'feed_type': 'json',
        'refresh_interval_secs': 600,
        'headers': {'User-Agent': 'test'}
    }
    
    feed = validate_feed(feed_data, 1)
    assert feed.name == 'Test Feed'
    assert feed.url == 'https://example.com/feed'
    assert feed.feed_type == 'json'
    assert feed.refresh_interval_secs == 600
    assert feed.headers == {'User-Agent': 'test'}


def test_validate_feed_missing_name():
    """Test feed validation with missing name"""
    feed_data = {
        'url': 'https://example.com',
        'feed_type': 'json'
    }
    
    with pytest.raises(ConfigError, match="missing field 'name'"):
        validate_feed(feed_data, 1)


def test_validate_feed_empty_name():
    """Test feed validation with empty name"""
    feed_data = {
        'name': '',
        'url': 'https://example.com',
        'feed_type': 'json'
    }
    
    with pytest.raises(ConfigError, match="non-empty string"):
        validate_feed(feed_data, 1)


def test_validate_feed_missing_url():
    """Test feed validation with missing URL"""
    feed_data = {
        'name': 'Test',
        'feed_type': 'json'
    }
    
    with pytest.raises(ConfigError, match="missing field 'url'"):
        validate_feed(feed_data, 1)


def test_validate_feed_invalid_url():
    """Test feed validation with invalid URL"""
    feed_data = {
        'name': 'Test',
        'url': 'not-a-url',
        'feed_type': 'json'
    }
    
    with pytest.raises(ConfigError, match="invalid URL"):
        validate_feed(feed_data, 1)


def test_validate_feed_missing_feed_type():
    """Test feed validation with missing feed_type"""
    feed_data = {
        'name': 'Test',
        'url': 'https://example.com'
    }
    
    with pytest.raises(ConfigError, match="missing field 'feed_type'"):
        validate_feed(feed_data, 1)


def test_validate_feed_invalid_feed_type():
    """Test feed validation with invalid feed_type"""
    feed_data = {
        'name': 'Test',
        'url': 'https://example.com',
        'feed_type': 'xml'
    }
    
    with pytest.raises(ConfigError, match="must be one of: json, rss, atom"):
        validate_feed(feed_data, 1)


def test_validate_feed_invalid_refresh_interval():
    """Test feed validation with invalid refresh interval"""
    feed_data = {
        'name': 'Test',
        'url': 'https://example.com',
        'feed_type': 'json',
        'refresh_interval_secs': -1
    }
    
    with pytest.raises(ConfigError, match="positive integer"):
        validate_feed(feed_data, 1)


def test_validate_settings_defaults():
    """Test settings validation with defaults"""
    settings = validate_settings({})
    assert settings.max_concurrency == 5
    assert settings.default_timeout_secs == 10
    assert settings.retry_max == 3
    assert settings.retry_base_delay_ms == 500
    assert settings.database_path == "feedpulse.db"


def test_validate_settings_custom():
    """Test settings validation with custom values"""
    settings_data = {
        'max_concurrency': 10,
        'default_timeout_secs': 30,
        'retry_max': 5,
        'retry_base_delay_ms': 1000,
        'database_path': 'custom.db'
    }
    
    settings = validate_settings(settings_data)
    assert settings.max_concurrency == 10
    assert settings.default_timeout_secs == 30
    assert settings.retry_max == 5
    assert settings.retry_base_delay_ms == 1000
    assert settings.database_path == 'custom.db'


def test_validate_settings_invalid_max_concurrency():
    """Test settings validation with invalid max_concurrency"""
    with pytest.raises(ConfigError, match="between 1 and 50"):
        validate_settings({'max_concurrency': 0})
    
    with pytest.raises(ConfigError, match="between 1 and 50"):
        validate_settings({'max_concurrency': 51})


def test_validate_settings_invalid_timeout():
    """Test settings validation with invalid timeout"""
    with pytest.raises(ConfigError, match="positive integer"):
        validate_settings({'default_timeout_secs': -1})

File: rust/tests/test_config.rs
/// Tests for configuration validation
use feedpulse::config::Config;
use std::fs;
use tempfile::NamedTempFile;

#[test]
fn test_config_missing_file() {
    let result = Config::load("/nonexistent/config.yaml");
    assert!(result.is_err());
    let err = result.unwrap_err();
    assert!(err.contains("Failed to read config file"), "Expected 'Failed to read config file', got: {}", err);
}

#[test]
fn test_config_invalid_yaml() {
    let mut temp_file = NamedTempFile::new().unwrap();
    fs::write(&temp_file, "invalid: yaml: content: [").unwrap();
    
    let result = Config::load(temp_file.path());
    assert!(result.is_err());
    let err = result.unwrap_err();
    assert!(err.contains("invalid config"), "Expected 'invalid config', got: {}", err);
}

#[test]
fn test_config_missing_required_field() {
    let mut temp_file = NamedTempFile::new().unwrap();
    fs::write(&temp_file, r#"
feeds:
  - name: "Test"
    feed_type: json
"#).unwrap();
    
    let config = Config::load(temp_file.path());
    assert!(config.is_err() || config.unwrap().validate().is_err());
}

#[test]
fn test_config_invalid_url() {
    let mut temp_file = NamedTempFile::new().unwrap();
    fs::write(&temp_file, r#"
feeds:
  - name: "Test"
    url: "not-a-valid-url"
    feed_type: json
"#).unwrap();
    
    let config = Config::load(temp_file.path()).unwrap();
    let result = config.validate();
    assert!(result.is_err());
    let err = result.unwrap_err();
    assert!(err.contains("invalid URL"), "Expected 'invalid URL', got: {}", err);
}

#[test]
fn test_validate_feed_empty_name() {
    let mut temp_file = NamedTempFile::new().unwrap();
    fs::write(&temp_file, r#"
feeds:
  - name: ""
    url: "https://example.com"
    feed_type: json
"#).unwrap();
    
    let config = Config::load(temp_file.path()).unwrap();
    let result = config.validate();
    assert!(result.is_err());
    let err = result.unwrap_err();
    assert!(err.contains("name cannot be empty"), "Expected 'name cannot be empty', got: {}", err);
}

#[test]
fn test_validate_feed_missing_feed_type() {
    let yaml = r#"
feeds:
  - name: "Test"
    url: "https://example.com"
"#;
    let mut temp_file = NamedTempFile::new().unwrap();
    fs::write(&temp_file, yaml).unwrap();
    
    // Should fail on deserialization since feed_type is required
    let result = Config::load(temp_file.path());
    assert!(result.is_err());
}

#[test]
fn test_validate_feed_invalid_feed_type() {
    let mut temp_file = NamedTempFile::new().unwrap();
    fs::write(&temp_file, r#"
feeds:
  - name: "Test"
    url: "https://example.com"
    feed_type: xml
"#).unwrap();
    
    let config = Config::load(temp_file.path()).unwrap();
    let result = config.validate();
    assert!(result.is_err());
    let err = result.unwrap_err();
    assert!(err.contains("must be one of: json, rss, atom"), "Expected feed_type error, got: {}", err);
}

#[test]
fn test_validate_settings_defaults() {
    let mut temp_file = NamedTempFile::new().unwrap();
    fs::write(&temp_file, r#"
feeds:
  - name: "Test"
    url: "https://example.com"
    feed_type: json
"#).unwrap();
    
    let config = Config::load(temp_file.path()).unwrap();
    assert_eq!(config.settings.max_concurrency, 5);
    assert_eq!(config.settings.default_timeout_secs, 10);
    assert_eq!(config.settings.retry_max, 3);
    assert_eq!(config.settings.retry_base_delay_ms, 500);
    assert_eq!(config.settings.database_path, "feedpulse.db");
}

#[test]
fn test_validate_settings_custom() {
    let mut temp_file = NamedTempFile::new().unwrap();
    fs::write(&temp_file, r#"
settings:
  max_concurrency: 10
  default_timeout_secs: 30
  retry_max: 5
  retry_base_delay_ms: 1000
  database_path: "custom.db"
feeds:
  - name: "Test"
    url: "https://example.com"
    feed_type: json
"#).unwrap();
    
    let config = Config::load(temp_file.path()).unwrap();
    assert_eq!(config.settings.max_concurrency, 10);
    assert_eq!(config.settings.default_timeout_secs, 30);
    assert_eq!(config.settings.retry_max, 5);
    assert_eq!(config.settings.retry_base_delay_ms, 1000);
    assert_eq!(config.settings.database_path, "custom.db");
}

#[test]
fn test_validate_settings_invalid_max_concurrency_zero() {
    let mut temp_file = NamedTempFile::new().unwrap();
    fs::write(&temp_file, r#"
settings:
  max_concurrency: 0
feeds:
  - name: "Test"
    url: "https://example.com"
    feed_type: json
"#).unwrap();
    
    let config = Config::load(temp_file.path()).unwrap();
    let result = config.validate();
    assert!(result.is_err());
    let err = result.unwrap_err();
    assert!(err.contains("between 1-50"), "Expected max_concurrency error, got: {}", err);
}

#[test]
fn test_validate_settings_invalid_max_concurrency_high() {
    let mut temp_file = NamedTempFile::new().unwrap();
    fs::write(&temp_file, r#"
settings:
  max_concurrency: 51
feeds:
  - name: "Test"
    url: "https://example.com"
    feed_type: json
"#).unwrap();
    
    let config = Config::load(temp_file.path()).unwrap();
    let result = config.validate();
    assert!(result.is_err());
    let err = result.unwrap_err();
    assert!(err.contains("between 1-50"), "Expected max_concurrency error, got: {}", err);
}

#[test]
fn test_validate_settings_invalid_timeout() {
    let mut temp_file = NamedTempFile::new().unwrap();
    fs::write(&temp_file, r#"
settings:
  default_timeout_secs: 0
feeds:
  - name: "Test"
    url: "https://example.com"
    feed_type: json
"#).unwrap();
    
    let config = Config::load(temp_file.path()).unwrap();
    let result = config.validate();
    assert!(result.is_err());
    let err = result.unwrap_err();
    assert!(err.contains("must be positive"), "Expected timeout error, got: {}", err);
}

#[test]
fn test_validate_feed_invalid_refresh_interval() {
    let mut temp_file = NamedTempFile::new().unwrap();
    fs::write(&temp_file, r#"
feeds:
  - name: "Test"
    url: "https://example.com"
    feed_type: json
    refresh_interval_secs: 0
"#).unwrap();
    
    let config = Config::load(temp_file.path()).unwrap();
    let result = config.validate();
    assert!(result.is_err());
    let err = result.unwrap_err();
    assert!(err.contains("must be positive"), "Expected refresh_interval error, got: {}", err);
}

#[test]
fn test_validate_url_formats() {
    // Valid URLs
    let valid_urls = vec![
        "https://example.com",
        "http://example.com",
        "https://example.com/path?query=1",
        "https://example.com:8080/path",
    ];
    
    for url in valid_urls {
        let yaml = format!(r#"
feeds:
  - name: "Test"
    url: "{}"
    feed_type: json
"#, url);
        let mut temp_file = NamedTempFile::new().unwrap();
        fs::write(&temp_file, yaml).unwrap();
        
        let config = Config::load(temp_file.path()).unwrap();
        assert!(config.validate().is_ok(), "Should accept valid URL: {}", url);
    }
}

#[test]
fn test_config_full_valid() {
    let mut temp_file = NamedTempFile::new().unwrap();
    fs::write(&temp_file, r#"
settings:
  max_concurrency: 5
  default_timeout_secs: 10
  retry_max: 3
  retry_base_delay_ms: 500
  database_path: "feedpulse.db"

feeds:
  - name: "Test Feed 1"
    url: "https://example.com/feed1"
    feed_type: json
    refresh_interval_secs: 300
    headers:
      User-Agent: "test"
  
  - name: "Test Feed 2"
    url: "https://example.com/feed2"
    feed_type: rss
"#).unwrap();
    
    let config = Config::load(temp_file.path()).unwrap();
    assert!(config.validate().is_ok());
    assert_eq!(config.feeds.len(), 2);
    assert_eq!(config.feeds[0].name, "Test Feed 1");
    assert_eq!(config.feeds[1].name, "Test Feed 2");
}

#[test]
fn test_load_from_test_fixtures() {
    // Test loading the valid config from test-fixtures
    let valid_config_path = "../test-fixtures/valid-config.yaml";
    if std::path::Path::new(valid_config_path).exists() {
        let config = Config::load(valid_config_path);
        assert!(config.is_ok(), "Should load valid config from test-fixtures");
    }
    
    // Test loading invalid configs
    let invalid_yaml_path = "../test-fixtures/invalid-config-bad-yaml.yaml";
    if std::path::Path::new(invalid_yaml_path).exists() {
        let config = Config::load(invalid_yaml_path);
        assert!(config.is_err(), "Should fail on bad YAML");
    }
    
    let missing_url_path = "../test-fixtures/invalid-config-missing-url.yaml";
    if std::path::Path::new(missing_url_path).exists() {
        let result = Config::load(missing_url_path);
        if let Ok(config) = result {
            assert!(config.validate().is_err(), "Should fail validation on missing URL");
        }
    }
}
